bson/vector.go

```diff
@@ -0,0 +1,271 @@
+// Copyright (C) MongoDB, Inc. 2024-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package bson
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math"
+)
+
+// VectorDType represents the Vector data type.
+type VectorDType byte
+
+// These constants are vector data types.
+const (
+	Int8Vector      VectorDType = 0x03
+	Float32Vector   VectorDType = 0x27
+	PackedBitVector VectorDType = 0x10
+)
+
+// Stringer of VectorDType
+func (vt VectorDType) String() string {
+	switch vt {
+	case Int8Vector:
+		return "int8"
+	case Float32Vector:
+		return "float32"
+	case PackedBitVector:
+		return "packed bit"
+	default:
+		return "invalid"
+	}
+}
+
+// These are vector conversion errors.
+var (
+	errInsufficientVectorData = errors.New("insufficient data")
+	errNonZeroVectorPadding   = errors.New("padding must be 0")
+	errVectorPaddingTooLarge  = errors.New("padding larger than 7")
+)
+
+type vectorTypeError struct {
+	Method string
+	Type   VectorDType
+}
+
+// Error implements the error interface.
+func (vte vectorTypeError) Error() string {
+	return "Call of " + vte.Method + " on " + vte.Type.String() + " vector"
```

Comment:
Optional: Consider a clearer error message. Also, error messages are conventionally lowercase.

```suggestion
	return "cannot call " + vte.Method + " on a type " + vte.Type.String() + " vector"
```